python/hsfs/core/vector_server.py

```diff
@@ -383,114 +405,124 @@ def _vector_result(self, entry, prepared_statement_objects):
-
-                statement_results = {}
-                serving_keys = self._serving_key_by_serving_index[
-                    prepared_statement_index
-                ]
-                serving_keys_all_fg += serving_keys
-                # Use prefix from prepare statement because prefix from serving key is collision adjusted.
-                prefix_features = [
-                    (self._prefix_by_serving_index[prepared_statement_index] or "")
-                    + sk.feature_name
-                    for sk in self._serving_key_by_serving_index[
-                        prepared_statement_index
-                    ]
-                ]
-                for row in result_proxy:
-                    row_dict = row._asdict()
-                    # can primary key be complex feature?
-                    result_dict = self.deserialize_complex_features(
-                        self._complex_features, row_dict
-                    )
-                    # note: should used serialized value
-                    # as it is from users' input
-                    statement_results[
-                        self._get_result_key(prefix_features, row_dict)
-                    ] = result_dict
-
-                # add partial results to the global results
-                for i, entry in enumerate(entries):
-                    batch_results[i].update(
-                        statement_results.get(
-                            self._get_result_key_serving_key(serving_keys, entry), {}
-                        )
+            )
+            bind_params = {"batch_ids": entry_values_tuples}
+            entry_values.append(bind_params)
+
+        # run all the prepared statements in parallel using aiomysql engine
+        loop = asyncio.get_event_loop()
+        parallel_results = loop.run_until_complete(
+            self._execute_prep_statements(
+                list(prepared_statement_objects.values()), entry_values
+            )
+        )
+
+        # construct the results
+        for count, prepared_statement_index in enumerate(prepared_statement_objects):
+            # prepared_statement_index include fg with label only
+            # But _serving_key_by_serving_index include the index when the join_index is 0 (left side)
+            if prepared_statement_index not in self._serving_key_by_serving_index:
+                continue
+
+            statement_results = {}
+            serving_keys = self._serving_key_by_serving_index[prepared_statement_index]
+            # Use prefix from prepare statement because prefix from serving key is collision adjusted.
+            prefix_features = [
+                (self._prefix_by_serving_index[prepared_statement_index] or "")
+                + sk.feature_name
+                for sk in self._serving_key_by_serving_index[prepared_statement_index]
+            ]
+            # iterate over results by index of the prepared statement
+            for row in parallel_results[count]:
```

Comment:
length of `prepared_statement_objects` can be longer than that of `parallel_results` because some prepare statement may not be executed.
```
            if prepared_statement_index not in self._serving_key_by_serving_index:
                continue
```